File: scala/src/main/scala/sodium/LazyCell.scala
package sodium

class LazyCell[A](event: Stream[A], _lazyInitValue: Option[Lazy[A]]) extends Cell[A](event, None) {

  this.lazyInitValue = _lazyInitValue

  override /*protected*/ def sampleNoTrans(): A = {
    if (currentValue.isEmpty && lazyInitValue.isDefined) {
      currentValue = Some(lazyInitValue.get.f.apply())
      lazyInitValue = None
    }
    currentValue.get
  }
}

File: java/src/main/java/nz/sodium/LazyCell.java
package nz.sodium;

class LazyCell<A> extends Cell<A> {
    LazyCell(final Stream<A> event, final Lazy<A> lazyInitValue) {
        super(event, null);
        this.lazyInitValue = lazyInitValue;
    }

    @Override
    A sampleNoTrans()
    {
        if (value == null && lazyInitValue != null) {
            value = lazyInitValue.get();
            lazyInitValue = null;
        }
        return value;
    }
}

